File: src/main/java/main/utils/locale/LocaleManager.java
package main.utils.locale;

import lombok.Getter;
import lombok.SneakyThrows;
import main.main.Robertify;
import main.utils.GeneralUtils;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.yaml.snakeyaml.Yaml;

import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LocaleManager {
    private final static Logger logger = LoggerFactory.getLogger(LocaleManager.class);
    private final static HashMap<Long, LocaleManager> localeManagers = new HashMap<>();
    private final static HashMap<RobertifyLocale, Map<String, String>> locales = new HashMap<>();

    @Getter
    private RobertifyLocale locale;
    private Map<String, String> localeFile;
    private final Guild guild;
    private final long guildID;

    private LocaleManager(Guild guild, RobertifyLocale locale) {
        locales.putIfAbsent(locale, retrieveLocaleFile(locale));
        this.guild = guild;
        this.guildID = guild.getIdLong();
        this.locale = locale;
        this.localeFile = locales.get(locale);
    }

    public static LocaleManager getLocaleManager(Guild guild) {
        if (!localeManagers.containsKey(guild.getIdLong()))
            localeManagers.put(guild.getIdLong(), new LocaleManager(guild, RobertifyLocale.ENGLISH));
        return localeManagers.get(guild.getIdLong());
    }


    public void setLocale(RobertifyLocale locale) {
        locales.putIfAbsent(locale, retrieveLocaleFile(locale));

        this.locale = locale;
        this.localeFile = retrieveLocaleFile();
        new LocaleConfig(guild).setLocale(locale);
    }

    public static void reloadLocales() {
        locales.clear();
        RobertifyLocale.getAvailableLanguages()
                .forEach(locale -> locales.put(locale, retrieveLocaleFile(locale)));
        Robertify.getShardManager().getGuildCache().forEach(guild -> {
            if (localeManagers.containsKey(guild.getIdLong())) {
                LocaleManager localeManager = localeManagers.get(guild.getIdLong());
                localeManager.setLocale(localeManager.locale);
            }
        });
    }

    public String getMessage(LocaleMessage message) {
        return localeFile.get(message.name().toLowerCase());
    }

    @SafeVarargs
    public final String getMessage(LocaleMessage message, Pair<String, String>... placeholders) {
        String msg = localeFile.get(message.name().toLowerCase());

        if (placeholders.length != 0)
            for (final var placeholder : placeholders)
                msg = msg.replaceAll(Pattern.quote(placeholder.getLeft()), Matcher.quoteReplacement(placeholder.getRight()));
        return msg;
    }

    @SneakyThrows
    private Map<String, String> retrieveLocaleFile() {
        return retrieveLocaleFile(locale);
    }

    @SneakyThrows
    private static Map<String, String> retrieveLocaleFile(RobertifyLocale locale) {
        if (!GeneralUtils.directoryExists("./locale")) {
            GeneralUtils.createDirectory("./locale");
            logger.error("There was no file found for locale: " + locale.getCode().toUpperCase());
            createAllLocaleFiles();
            System.exit(-1);
        }

        File localeFile = new File("./locale/messages." + locale.getCode().toLowerCase() + ".yml");

        if (!localeFile.exists())
            createLocaleFile(locale);

//        if (localeFile.length() == 0) {
//            logger.error("There was no information found in the file for locale: " + locale.getCode().toUpperCase());
//            System.exit(-1);
//        }

        Map<String, String> fileMap = new Yaml().load(GeneralUtils.getFileContent(localeFile.getPath()));
        return checkLocaleFile(localeFile, fileMap);
    }

    @SneakyThrows
    private void createLocaleFile() {
        createLocaleFile(locale);
    }

    @SneakyThrows
    private static void createLocaleFile(RobertifyLocale locale) {
        final File file = new File("./locale/messages." + locale.getCode().toLowerCase() + ".yml");
        file.createNewFile();

        StringBuilder content = new StringBuilder();
        for (final var fieldSection : RobertifyLocaleMessage.getMessageTypes().values())
            for (final var field : fieldSection)
                content.append(field.name().toLowerCase()).append(": ").append("\"Fill me out!\"\n");
        GeneralUtils.setFileContent(file, content.toString());
    }

    private static void createAllLocaleFiles() {
        for (final var locale : RobertifyLocale.values())
            createLocaleFile(locale);
    }

    private static Map<String, String> checkLocaleFile(File localeFile, Map<String, String> fileMap) {
        StringBuilder contentToAppend = new StringBuilder();
        for (final var fieldSection : RobertifyLocaleMessage.getMessageTypes().values()) {
            for (final var field : fieldSection)
                if (!fileMap.containsKey(field.name().toLowerCase())) {
                    contentToAppend.append(field.name().toLowerCase()).append(": \"Fill me out!\"\n");
                    fileMap.put(field.name().toLowerCase(), "Fill me out!");
                }
        }

        GeneralUtils.appendFileContent(localeFile, contentToAppend.toString());
        return fileMap;
    }
}

File: src/main/kotlin/main/utils/locale/LocaleManager.kt
package main.utils.locale

import kotlinx.coroutines.runBlocking
import main.utils.GeneralUtils
import main.utils.GeneralUtils.setContent
import main.utils.json.locale.LocaleConfig
import main.utils.locale.messages.RobertifyLocaleMessageKt.Companion.getMessageTypes
import net.dv8tion.jda.api.entities.Guild
import org.slf4j.LoggerFactory
import org.yaml.snakeyaml.Yaml
import java.io.File
import java.io.FileInputStream
import java.util.*
import kotlin.system.exitProcess

class LocaleManager private constructor(private val guild: Guild?, _locale: RobertifyLocale) {

    companion object {
        private val log = LoggerFactory.getLogger(Companion::class.java)
        private val localeManagers = Collections.synchronizedMap(emptyMap<Long, LocaleManager>().toMutableMap())
        private val locales = Collections.synchronizedMap(emptyMap<RobertifyLocale, Map<String, String>>().toMutableMap())

        fun globalManager(): LocaleManager = LocaleManager(null, RobertifyLocale.ENGLISH)

        fun getLocaleManager(guild: Guild?): LocaleManager {
            if (guild == null)
                return globalManager()
            return localeManagers.computeIfAbsent(guild.idLong) {
                LocaleManager(guild, RobertifyLocale.ENGLISH)
            }
        }

        operator fun get(guild: Guild?): LocaleManager {
            if (guild == null)
                return globalManager()
            return localeManagers.computeIfAbsent(guild.idLong) {
                LocaleManager(guild, RobertifyLocale.ENGLISH)
            }
        }

        private fun retrieveLocaleFile(locale: RobertifyLocale): Map<String, String> {
            if (!GeneralUtils.directoryExists("./locale")) {
                GeneralUtils.createDirectory("./locale")
                log.error("There was no file found for locale: ${locale.code.uppercase()}")
                createAllLocaleFiles()
                exitProcess(-1)
            }

            val localeFile = File("./locale/messages.${locale.code.lowercase()}.yml")
            if (!localeFile.exists()) {
                log.warn("${locale.code} locale didn't exist. Creating it.")
                createLocaleFile(locale)
            }

            return Yaml().load(FileInputStream(localeFile))
        }

        private fun createLocaleFile(locale: RobertifyLocale) {
            val file = File("./locale/messages.${locale.code.lowercase()}.yml")
            val content = StringBuilder()

            for (fieldSection in getMessageTypes().values)
                for (field in fieldSection)
                    content.append(field.name.lowercase())
                        .append(": ")
                        .append("\"Fill me out\"")
            file.setContent(content.toString())
        }

        private fun createAllLocaleFiles() =
            RobertifyLocale.values().forEach { createLocaleFile(it) }

        fun reloadLocales() {
            locales.clear()
            RobertifyLocale.availableLanguages
                .forEach { locale -> locales[locale] = retrieveLocaleFile(locale) }
        }

    }

    fun getLocale(): RobertifyLocale {
        if (guild == null)
            return RobertifyLocale.ENGLISH
        return LocaleConfig(guild).getLocale()
    }

    fun setLocale(locale: RobertifyLocale) {
        locales.putIfAbsent(locale, retrieveLocaleFile(locale))

        if (guild != null)
            LocaleConfig(guild).setLocale(locale)
    }

    fun getLocaleFile(): Map<String, String> {
        return locales.computeIfAbsent(getLocale()) {
            retrieveLocaleFile(it)
        }
    }

    fun getMessage(message: LocaleMessage, vararg placeholders: Pair<String, String>): String {
        var msg = getLocaleFile()[message.name.lowercase()]
            ?: throw NullPointerException("There was no such message found in the mapping with key: ${message.name}")

        placeholders.forEach { placeholder ->
            msg = msg.replace(placeholder.first, placeholder.second)
        }

        return msg
    }
}